### PR TITLE
[intro.races] Clarify conflicts for the case where no bits are changed

### LINKED ISSUE #6551: [intro.races] It is slightly unclear whether a data race can occur if no bits are changed
See OP's confusion in https://stackoverflow.com/q/77031819/5740428.

I believe that the current wording implies the following (I've elaborated on this in my answer https://stackoverflow.com/a/77032028/5740428):
```cpp
int x = 0; // global variable

void concurrently() {
    x = 0; // may be a data race, even if the value of x prior to this assignment is zero
}
```
I think there should be a note or example somewhere (possibly [defns.access], [intro.races], [intro.multithread]) which explains that even if the value is unchanged (based on equality comparison), changing the value (in the abstract machine sense) still counts as modification.

### PR DESCRIPTION
**Pull request by Eisenwave**
Fixes #6551.

### CODE REVIEW

File: source/expressions.tex [hunk 1]
```diff
@@ -7310,6 +7310,10 @@
 In simple assignment (\tcode{=}), the object referred to by the left operand
 is modified\iref{defns.access}
 by replacing its value with the result of the right operand.
+\begin{note}
+The object is modified even if the value of the right operand
+compares equal\iref{expr.eq} to the object's old value.
```
Comment (CaseyCarter) [inline source/expressions.tex:7315]: I don't think this redundancy makes "the object...is modified" any clearer. If anything, it makes me think I must have missed some subtlety in the normative wording.

Comment (Eisenwave) [reply source/expressions.tex:7315]: How would you phrase it then? The point this is meant to get across is that even if `x` is `0`, `x = 0` is considered to be modification. This is not obvious from the term "modify" itself, since it is defined as "make partial or minor changes to (something)" in English.

Comment (jensmaurer) [reply source/expressions.tex:7315]: "modify" comes with a cross-reference after it, so readers are aware this is a technical term that has more than the English meaning.  If you want to add a cross-reference to the data races section, I'm fine with it.  However, there is no hint in the sentence that the "modify" would be somehow conditional on the values involved, and the "replace" sentence is pretty straightforward, so I'm not seeing a need for a note.

Again, if you wish to suggest adding more cross-reference to add more load to the word "modify", go for it.

Comment (Eisenwave) [reply source/expressions.tex]: Ok, so admittedly, [expr.ass] was probably not the right place to approach it. The PR now tackles the problem in [intro.races].

**Comment by Eisenwave** [comment]
The new approach is now to
- add a cross-reference to [defns.access], and
- add a note which clarifies in [intro.races] that lack of bit changes does not absolve a modification from conflicting with another.

It is sufficiently clear that [expr.ass] always performs a modification, but due to the nature of the term "modify", it is not obvious that a modification that doesn't actually modify any bits can cause a data race.

File: source/basic.tex [hunk 2]
```diff
@@ -6169,9 +6169,15 @@
...
 \pnum
-Two expression evaluations \defn{conflict} if one of them modifies a memory
+Two expression evaluations \defn{conflict} if one of them modifies\iref{defns.access} a memory
 location\iref{intro.memory} and the other one reads or modifies the same
 memory location.
+\begin{note}
+A modification conflicts with another expression evaluation
+at the same memory location
+even if the modification alters no bits in the object representation
+of the modified memory location.
```
Comment (jensmaurer) [inline source/basic.tex:6179]: A memory location is not necessarily an "object", thus the term "object representation" might not be defined for it.

Suggestion for abbreviation: "Such conflicts occur even if a participating modification does not alter any bits."

Comment (Eisenwave) [reply source/basic.tex]: I agree that the old proposed wording was a bit too lengthy and still didn't achieve correctness.

The problem with *"Such conflicts"* is that we have defined what it means *to conflict* (verb), so ideally, we would not use the noun counterpart since it's ever so slightly informal.

I've settled on
> When a modification alters no bits, this does not prevent if from conflicting.

File: source/basic.tex [hunk 3]
```diff
@@ -6169,9 +6169,13 @@
 \end{note}
 
 \pnum
-Two expression evaluations \defn{conflict} if one of them modifies a memory
+Two expression evaluations \defn{conflict} if one of them modifies\iref{defns.access} a memory
 location\iref{intro.memory} and the other one reads or modifies the same
 memory location.
+\begin{note}
+When a modification alters no bits,
+this does not prevent if from conflicting.
```
Comment (jwakely) [inline source/basic.tex:6177]:
```suggestion
this does not prevent it from conflicting.
```

Comment (jwakely) [reply source/basic.tex:6177]: How about:

> A modification can still conflict even if it does not alter the value of any bits.

I'm a little unhappy about just saying "can conflict" rather than "can conflict with another expression evaluation" but that starts to get too wordy again.

An even wordier form of the same sentence would be:

> A modification can still conflict with another expression evaluation even if it does not alter the value of any bits in the memory location.

Comment (Eisenwave) [reply source/basic.tex]: I think
> A modification can still conflict even if it does not alter the value of any bits.

is the best option out of these. Since "to conflict" has only been defined for expression evaluations, it is immediately obvious what the modification can conflict with, so we don't need to be so explicit.

**Review by jwakely — APPROVED** [review]
I'm not a Core wording or concurrency wording expert, but I like this.

**Review by jensmaurer — APPROVED** [review]
